book/src/developers/designs/replication.md

```diff
@@ -0,0 +1,465 @@
+The tombstone propagates between all nodes of the topology, and after a tombstone window has passed,
+is reaped by all nodes internally.
+
+A live entry also contains a map of change times. This contains the maximum CID of when an attribute
+of the entry was updated last. Consider an entry like:
+
+    attr_a: 1
+    attr_b: 2
+    attr_c: 3
+    uuid:   X
+
+This entries changestate would show:
+
+    Live {
+      at: { server_uuid: A, cid: 1 },
+      attrs: {
+        attr_a: cid = 1
+        attr_b: cid = 1
+        attr_c: cid = 2
+      }
+    }
+
+This shows us that the entry was created on server A, at cid time 1. At creation, the attributes a
+and b were created since they have the same cid.
+
+attr c was either updated or created after this - we can't tell if it existed at cid 1, we can only
+know that a write of some kind occurred at cid 2.
+
+## Resolving Conflicts
+
+With knowledge of the change state structure we can now demonstrate how the lower level entry and
+attribute conflicts are detected and managed in Kanidm.
+
+### Entry
+
+An entry conflict occurs when two servers create an entry with the same UUID at the same time. This
+would be shown as:
+
+            Server A            Server B
+    Time 0: create entry X
+    Time 1:                     create entry X
+    Time 2:       <-- incremental --
+    Time 3:        -- incremental -->
+
+We can add in our entry change state for liveness here.
+
+    Time 0: create entry X cid { time: 0, server: A }
+    Time 1:                     create entry X cid { time: 1, server: B }
+    Time 2:       <-- incremental --
+    Time 3:        -- incremental -->
+
+When the incremental occurs at time point 2, server A would consider these on a timeline as:
+
+    Time 0: create entry X cid { time: 0, server: A }
+    Time 1: create entry X cid { time: 1, server: B }
+
+When viewed like this, we can see that the second create had it been performed on the same server
+would have been rejected for a duplicate entry. In replication though, this means that the later
+entry will be moved to the conflict state.
+
+The same process occurs with the same results when the reverse incremental operation occurs to
+server B where it receives the entry with the earlier creation from A. It will order the events and
+conflict it's local copy of the entry.
+
+### Attribute
+
+An attribute conflict occurs when two servers modify the same attribute of the same entry before an
+incremental replication occurs.
+
+            Server A            Server B
+    Time 0: create entry X
+    Time 1:        -- incremental -->
+    Time 2: modify entry X
+    Time 3:                     modify entry X
+    Time 4:       <-- incremental --
+    Time 5:        -- incremental -->
+
+During an incremental operation, a modification to a live entry is allowed to apply provided the
+entries uuid and at match the servers metadata. This gives the servers assurance that an entry is
```

Comment:
what's the "at" - attribute?

```suggestion
entry's UUID and at match the servers metadata. This gives the servers assurance that an entry is
```

Reply:
created *at*. It's a terrible name. I regret it.